test/resource/validation/compare_test.exs

```diff
@@ -32,6 +32,396 @@ defmodule Ash.Test.Resource.Validation.CompareTest do
+
+      assert :ok = Compare.validate(changeset, opts, %{})
+
+      changeset = Post |> Ash.Changeset.for_create(:create, %{list_value: [4, 5, 6]})
+
+      assert_error(changeset, opts, "must be equal to [1, 2, 3]")
+    end
+
+    test "maps can be compared" do
+      {:ok, opts} = Compare.init(attribute: :map_value, is_equal: %{a: 1, b: 2})
+      changeset = Post |> Ash.Changeset.for_create(:create, %{map_value: %{a: 1, b: 2}})
+
+      assert :ok = Compare.validate(changeset, opts, %{})
+
+      changeset = Post |> Ash.Changeset.for_create(:create, %{map_value: %{a: 3, b: 4}})
+
+      assert_error(changeset, opts, "must be equal to %{a: 1, b: 2}")
+    end
+
+    test "tuples can be compared" do
+      {:ok, opts} = Compare.init(attribute: :tuple_value, is_equal: {1, 2})
+      changeset = Post |> Ash.Changeset.for_create(:create, %{tuple_value: {1, 2}})
+
+      assert :ok = Compare.validate(changeset, opts, %{})
+
+      changeset = Post |> Ash.Changeset.for_create(:create, %{tuple_value: {3, 4}})
+
+      assert_error(changeset, opts, "must be equal to {1, 2}")
+    end
+
+    test "atoms can be compared" do
+      # Use a tuple to represent a literal value
+      {:ok, opts} = Compare.init(attribute: :atom_value, is_equal: {:literal, :foo})
+      changeset = Post |> Ash.Changeset.for_create(:create, %{atom_value: :foo})
+
+      assert :ok = Compare.validate(changeset, opts, %{})
+
+      changeset = Post |> Ash.Changeset.for_create(:create, %{atom_value: :bar})
+
+      assert_error(changeset, opts, "must be equal to foo")
+    end
+
+    test "strings can be compared" do
+      {:ok, opts} = Compare.init(attribute: :string_value, is_equal: "hello")
+      changeset = Post |> Ash.Changeset.for_create(:create, %{string_value: "hello"})
+
+      assert :ok = Compare.validate(changeset, opts, %{})
+
+      changeset = Post |> Ash.Changeset.for_create(:create, %{string_value: "world"})
+
+      assert_error(changeset, opts, "must be equal to hello")
+    end
+
+    test "booleans can be compared" do
+      # Use a tuple to represent a literal value
+      {:ok, opts} = Compare.init(attribute: :boolean_value, is_equal: {:literal, true})
+      changeset = Post |> Ash.Changeset.for_create(:create, %{boolean_value: true})
+
+      assert :ok = Compare.validate(changeset, opts, %{})
+
+      changeset = Post |> Ash.Changeset.for_create(:create, %{boolean_value: false})
+
+      assert_error(changeset, opts, "must be equal to true")
+    end
+
+    test "nil can be compared" do
+      # Use a tuple to represent a literal nil value
+      {:ok, opts} = Compare.init(attribute: :nil_value, is_equal: {:literal, nil})
+      changeset = Post |> Ash.Changeset.for_create(:create, %{nil_value: nil})
+
+      assert :ok = Compare.validate(changeset, opts, %{})
+
+      changeset = Post |> Ash.Changeset.for_create(:create, %{nil_value: "not nil"})
+
+      assert_error(changeset, opts, "must be equal to ")
```

Comment:
should we do some extra special casing for this scenario? I guess you'd want to see "must be equal to nil"

Reply:
We probably should not allow `equal_to: nil`, but should instead require `is_nil: true/false`. Its unlikely in any other case you'd use `nil`, (because comparing to `nil` is weird in these contexts) but we should probably forbid it as a value in any of these options.